src/Quaternion.jl

```diff
@@ -110,15 +110,58 @@ end
 
 argq(q::Quaternion) = normalizeq(Quaternion(0, q.v1, q.v2, q.v3))
 
-function exp(q::Quaternion)
+# extend complex analytic function f to the quaternion q
+# adapted from Theorem 5 of doi.org/10.1017/S0305004100055638
+function extend_analytic(f, q::Quaternion)
+    a = abs_imag(q)
     s = q.s
-    se = exp(s)
-    scale = se
-    th = abs_imag(q)
-    if th > 0
-        scale *= sin(th) / th
+    z = complex(s, a)
+    w = f(z)
+    wr, wi = reim(w)
+    scale = wi / a
+    norm = _isexpfun(f) && abs(s) < eps(float(typeof(s)))
+    if a > 0
+        return Quaternion(wr, scale * q.v1, scale * q.v2, scale * q.v3, norm)
+    else  # quaternion may be real or complex
+        return Quaternion(wr, oftype(scale, wi), zero(scale), zero(scale), norm)
+    end
+end
+
+_isexpfun(::Union{typeof(exp),typeof(exp2),typeof(exp10)}) = true
+_isexpfun(::Any) = false
+
+for f in (
+    :sqrt, :exp, :exp2, :exp10, :expm1, :log2, :log10, :log1p,
+    :sin, :cos, :tan, :asin, :acos, :atan, :sinh, :cosh, :tanh, :asinh, :acosh, :atanh,
+    :csc, :sec, :cot, :acsc, :asec, :acot, :csch, :sech, :coth, :acsch, :asech, :acoth,
+    :sinpi, :cospi,
+)
+    @eval Base.$f(q::Quaternion) = extend_analytic($f, q)
+end
+
+for f in (@static(VERSION ≥ v"1.6" ? (:sincos, :sincospi) : (:sincos,)))
```

Comment:
I think we can drop support under v1.6 because we have v1.6 LTS. (`Project.toml` will also needed to be updated)
```suggestion
for f in (:sincos, :sincospi)
```

Reply:
We could, but IMO this should be handled outside this PR. Also, so far maintaining support for Julia 1.0 hasn't been an issue, so I don't think we need to rush dropping pre-1.6 support.